### ISSUE TITLE
pgbackrest dedicated repo server with GCS. Stanza created but backup failed.

### ISSUE THREAD

**Issue by vhadianto** [pgbackrest/pgbackrest#1776]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version: 2.39

2. PostgreSQL version: 14.2

3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each: Debian Bullseye

4. Did you install pgBackRest from source or from a package? Package

Setup:

1. Server: Postgres & pgbackrest running in TLS server mode
2. Remo: pgbackrest running in TLS server mode

**Postgres conf:**

```
    archive_command = 'pgbackrest --config=/opt/pgbackrest/server/config/pgbackrest.conf --stanza=demo archive-push %p'
    archive_mode = on
    # listen_addresses = '*'
    log_line_prefix = ''
    max_wal_senders = 3
    wal_level = replica
```

**Server's pgbackrest.conf:**

```
    [global]
    repo1-host=repo
    repo1-host-user=pgbackrest
    repo1-host-type=tls
    repo1-host-cert-file=/opt/pgbackrest/server/certs/server.crt
    repo1-host-key-file=/scratch/server.key
    repo1-host-ca-file=/opt/pgbackrest/server/certs/ca.crt

    # general options
    process-max=2
    log-level-console=debug
    log-level-file=debug
    log-path=/scratch

    # tls server options
    tls-server-address=*
    tls-server-cert-file=/opt/pgbackrest/server/certs/server.crt
    tls-server-key-file=/scratch/server.key
    tls-server-ca-file=/opt/pgbackrest/server/certs/ca.crt
    tls-server-auth=repo=demo

    [demo]
    pg1-path=/home/pg/db/14.2.0/data
```

**Repo server's pgbackrest.conf:**
```
    [global]
    repo1-type=gcs
    repo1-path=/backup
    repo1-gcs-key=/opt/pgbackrest/serviceaccount/gcs-key.json
    repo1-gcs-bucket=vh-backup-bucket

    repo1-retention-full=2

    # general options
    process-max=2
    # log-level-console=info
    # log-level-file=info

    # change this
    log-path=/scratch
    start-fast=y
    delta=y

    # tls server options
    tls-server-address=*
    tls-server-cert-file=/opt/pgbackrest/repo/certs/repo.crt
    tls-server-key-file=/scratch/repo.key
    tls-server-ca-file=/opt/pgbackrest/repo/certs/ca.crt
    tls-server-auth=pgserver=demo

    [demo]
    pg1-host=pgserver
    pg1-path=/home/pg/db/14.2.0/data
    pg1-host-type=tls
    pg1-host-cert-file=/opt/pgbackrest/repo/certs/repo.crt
    pg1-host-key-file=/scratch/repo.key
    pg1-host-ca-file=/opt/pgbackrest/repo/certs/ca.crt
    pg1-database=rre5xr
    pg1-socket-path=/tmp
    pg1-user=root
```

**Issue:**

I can create the stanza using the `stanza-create` command. I checked GCS bucket and I could see that the `archive.info` file has been created. However, I am unable actually do the backup.

When I run the `check` command in the `repo` server:

```sh
pgbackrest --config=/opt/pgbackrest/repo/config/pgbackrest.conf --log-level-console=debug --stanza=demo check
```

I saw the following error in the Postgres log:
```
2022-06-09 01:07:43.063 P00  DEBUG:     common/exit::exitSafe: (result: 0, error: true, signalType: 0)
ERROR: [103]: unable to find a valid repository:
       repo1: [FileOpenError] unable to load info file '/var/lib/pgbackrest/archive/demo/archive.info' or '/var/lib/pgbackrest/archive/demo/archive.info.copy':
       FileOpenError: raised from remote-0 tls protocol on 'repo': unable to open file '/var/lib/pgbackrest/archive/demo/archive.info' for read: [13] Permission denied
       FileOpenError: raised from remote-0 tls protocol on 'repo': unable to open file '/var/lib/pgbackrest/archive/demo/archive.info.copy' for read: [13] Permission denied
       HINT: archive.info cannot be opened but is required to push/get WAL segments.
       HINT: is archive_command configured correctly in postgresql.conf?
       HINT: has a stanza-create been performed?
       HINT: use --no-archive-check to disable archive checks during backup if you have an alternate archiving scheme.
```

As you can see the `remote-0` server is checking the file system for the `archive.info` file while it should be checking the GCS bucket.

I tried various setting, i.e. moving the `repo1-gcs*` settings in the `[demo]` sections in one and then both of the pgbackrest.conf file. 

If I added the `gcs` setting in the `[global]` section in the `server's pgbackrest.conf` it will complain that it needs `repo1-gcs-key` which exists.

Any one has any idea how to configure this correctly? Appreciate any help/hint/suggestions.

Thanks.

PS: I can post here the various settings that I have tried but thought (for brevity) that I should post to the setting that I think should work and start from there.

**Comment by dwsteele** [comment]
This has something to do with not being able to find the right config, but I'm not immediately seeing it. Putting all your configs in a non-standard and unique location per server is going to cause a ton of confusion.

Add `--log-level-console=info` to archive_command so we can see what config it is loading. Add the same option when you run `check` and provide the output.

**Comment by vhadianto** [comment]
Hi @dwsteele ,

**tl;dr;** In this TLS server setup, the archive command failed if the config is not located in `/etc/pgbackrest.conf`


Thanks for responding. I have further info on this issue, but first, to address the reason why the config files are located in "non-standard and unique" location. I am working on a dynamic Postgres database provisioning in K8s, we want our base Postgres container to be read only and we mount dynamically generated config files into different directories in the running container. This is also the reason why Postgres' socket is in `/tmp`

I got your point that it's difficult to debug this issue when things are located in unusual locations. I do have a simple setup in my local computer with 2 Docker containers that I was using to initially set this up. So I went back to this and set it up for GCS repo.

I now have the following config.

**Repo server**
```
[global]

repo1-type=gcs
repo1-path=/localhost
repo1-gcs-key=/home/pgbackrest/certs/gcs-key.json
repo1-gcs-bucket=vh-steampipe-workspace-backup-apse1


# repo details
#repo1-path=/backup_space
repo1-retention-full=2

# general options
process-max=2
log-level-console=info
log-level-file=debug
start-fast=y
delta=y

# tls server options
tls-server-address=*
tls-server-cert-file=/home/pgbackrest/certs/backup-srv.crt
tls-server-key-file=/home/pgbackrest/certs/backup-srv.key
tls-server-ca-file=/home/pgbackrest/certs/ca.crt
tls-server-auth=postgresql=demo

[demo]
pg1-host=postgresql
pg1-path=/var/lib/pgsql/14/data
pg1-host-type=tls
pg1-host-cert-file=/home/pgbackrest/certs/backup-srv.crt
pg1-host-key-file=/home/pgbackrest/certs/backup-srv.key
pg1-host-ca-file=/home/pgbackrest/certs/ca.crt
```

**Postgres server**
```
[global]
repo1-host=pgbackrest
repo1-host-user=pgbackrest
repo1-host-type=tls
repo1-host-cert-file=/var/lib/pgsql/certs/pg1-srv.crt
repo1-host-key-file=/var/lib/pgsql/certs/pg1-srv.key
repo1-host-ca-file=/var/lib/pgsql/certs/ca.crt

# general options
process-max=2
log-level-console=info
log-level-file=debug

# tls server options
tls-server-address=*
tls-server-cert-file=/var/lib/pgsql/certs/pg1-srv.crt
tls-server-key-file=/var/lib/pgsql/certs/pg1-srv.key
tls-server-ca-file=/var/lib/pgsql/certs/ca.crt
tls-server-auth=pgbackrest=demo

[demo]
pg1-path=/var/lib/pgsql/14/data
```

And my archive command is simply: `archive_command = 'pgbackrest --stanza=demo --log-level-console=info   archive-push %p'`

This works perfectly and as expected. I can see my repo was created in GCS and the subsequent backups.

However, when I tried simulating what will happen in our Kubernetes deployment I have the same exact issue. So this is what I did for the config files:

1. First move the config file to tmp dir: `mv /etc/pgbackrest.conf /tmp/pbackrest.conf`
2. Modify the archive command to: `pgbackrest --stanza=demo --config=/tmp/pgbackrest.conf --log-level-console=info   archive-push %p`
3. Restart Postgres, restart pgbackrest to read the config file from the /tmp directory, i.e. `pgbackrest --config=/tmp/pgbackrest.conf`

Now I can see the same issue in my k8s deployment:
1. `stanza-create` command is working fine, but
2. `check` command failed.

**stanza create**
```
pgbackrest --config=/tmp/pgbackrest.conf --log-level-console=info --stanza=demo stanza-create
2022-06-14 02:39:09.875 P00   INFO: stanza-create command begin 2.39: --config=/tmp/pgbackrest.conf --exec-id=183-f02352ca --log-level-console=info --log-level-file=debug --pg1-host=postgresql --pg1-host-ca-file=/home/pgbackrest/certs/ca.crt --pg1-host-cert-file=/home/pgbackrest/certs/backup-srv.crt --pg1-host-key-file=/home/pgbackrest/certs/backup-srv.key --pg1-host-type=tls --pg1-path=/var/lib/pgsql/14/data --repo1-gcs-bucket=vh-steampipe-workspace-backup-apse1 --repo1-gcs-key=<redacted> --repo1-path=/localhost --repo1-type=gcs --stanza=demo
2022-06-14 02:39:10.502 P00   INFO: stanza-create for stanza 'demo' on repo1
2022-06-14 02:39:11.101 P00   INFO: stanza 'demo' already exists on repo1 and is valid
2022-06-14 02:39:11.101 P00   INFO: stanza-create command end: completed successfully (1227ms)
```

**check**
```
pgbackrest --config=/tmp/pgbackrest.conf --log-level-console=info --stanza=demo check
2022-06-14 02:39:40.348 P00   INFO: check command begin 2.39: --config=/tmp/pgbackrest.conf --exec-id=185-99212994 --log-level-console=info --log-level-file=debug --pg1-host=postgresql --pg1-host-ca-file=/home/pgbackrest/certs/ca.crt --pg1-host-cert-file=/home/pgbackrest/certs/backup-srv.crt --pg1-host-key-file=/home/pgbackrest/certs/backup-srv.key --pg1-host-type=tls --pg1-path=/var/lib/pgsql/14/data --repo1-gcs-bucket=vh-steampipe-workspace-backup-apse1 --repo1-gcs-key=<redacted> --repo1-path=/localhost --repo1-type=gcs --stanza=demo
2022-06-14 02:39:40.990 P00   INFO: check repo1 configuration (primary)
2022-06-14 02:39:41.693 P00   INFO: check repo1 archive for WAL (primary)
```

**postgres log**
```
2022-06-14 02:39:42.679 P00   INFO: archive-push command end: aborted with exception [103]
2022-06-14 02:39:42.680 UTC [176] LOG:  archive command failed with exit code 103
2022-06-14 02:39:42.680 UTC [176] DETAIL:  The failed archive command was: pgbackrest --stanza=demo --config=/tmp/pgbackrest.conf --log-level-console=info   archive-push pg_wal/000000010000000000000017
2022-06-14 02:39:43.691 P00   INFO: archive-push command begin 2.39: [pg_wal/000000010000000000000017] --config=/tmp/pgbackrest.conf --exec-id=251-c655a36c --log-level-console=info --log-level-file=debug --pg1-path=/var/lib/pgsql/14/data --process-max=2 --repo1-host=pgbackrest --repo1-host-ca-file=/var/lib/pgsql/certs/ca.crt --repo1-host-cert-file=/var/lib/pgsql/certs/pg1-srv.crt --repo1-host-key-file=/var/lib/pgsql/certs/pg1-srv.key --repo1-host-type=tls --repo1-host-user=pgbackrest --stanza=demo
ERROR: [103]: unable to find a valid repository:
       repo1: [FileOpenError] unable to load info file '/var/lib/pgbackrest/archive/demo/archive.info' or '/var/lib/pgbackrest/archive/demo/archive.info.copy':
       FileOpenError: raised from remote-0 tls protocol on 'pgbackrest': unable to open file '/var/lib/pgbackrest/archive/demo/archive.info' for read: [13] Permission denied
       FileOpenError: raised from remote-0 tls protocol on 'pgbackrest': unable to open file '/var/lib/pgbackrest/archive/demo/archive.info.copy' for read: [13] Permission denied
       HINT: archive.info cannot be opened but is required to push/get WAL segments.
       HINT: is archive_command configured correctly in postgresql.conf?
       HINT: has a stanza-create been performed?
       HINT: use --no-archive-check to disable archive checks during backup if you have an alternate archiving scheme.
2022-06-14 02:39:43.705 P00   INFO: archive-push command end: aborted with exception [103]
2022-06-14 02:39:43.706 UTC [176] LOG:  archive command failed with exit code 103
2022-06-14 02:39:43.706 UTC [176] DETAIL:  The failed archive command was: pgbackrest --stanza=demo --config=/tmp/pgbackrest.conf --log-level-console=info   archive-push pg_wal/000000010000000000000017
2022-06-14 02:39:43.706 UTC [176] WARNING:  archiving write-ahead log file "000000010000000000000017" failed too many times, will try again later
```

I reverted everything back to "standard location" again:
1. config files back in /etc/pgbackrest.conf
2. fixed `archive_command`

And it's working again.

My current theory is that in this TLS server setup, the archive command failed *if* the config is *not* located in /etc/pgbackrest.conf

FYI, I've also attached my docker-compose.yaml file if that helps.

[docker-compose.yml.zip](https://github.com/pgbackrest/pgbackrest/files/8895925/docker-compose.yml.zip)

**Comment by dwsteele** [comment]
OK, I see it now. You have not set `repo-host-config` (https://pgbackrest.org/configuration.html#section-repository/option-repo-host-config) and `pg-host-config` (https://pgbackrest.org/configuration.html#section-stanza/option-pg-host-config). 

The archive-push command does not know where the find the config file on the repo-host, so it tries to use defaults. You can put these options in the pg/repo host config files since you are telling archive-push where the local config file is.

**Comment by vhadianto** [comment]
Thanks for your help @dwsteele I got it to work. I just need to add the `repo1-host-config` setting in the Postgres' pgbackrest.conf.

I have my working config here for the benefit of posterity.

**Repo server**
```
[global]
repo1-type=gcs
repo1-path=/localhost
repo1-gcs-key=/home/pgbackrest/certs/gcs-key.json
repo1-gcs-bucket=vh-backup-bucket

# repo details
repo1-retention-full=2

# general options
process-max=2
log-level-console=info
log-level-file=debug
start-fast=y
delta=y

# tls server options
tls-server-address=*
tls-server-cert-file=/home/pgbackrest/certs/backup-srv.crt
tls-server-key-file=/home/pgbackrest/certs/backup-srv.key
tls-server-ca-file=/home/pgbackrest/certs/ca.crt
tls-server-auth=postgresql=demo

[demo]
pg1-host=postgresql
pg1-path=/var/lib/pgsql/14/data
pg1-host-type=tls
pg1-host-cert-file=/home/pgbackrest/certs/backup-srv.crt
pg1-host-key-file=/home/pgbackrest/certs/backup-srv.key
pg1-host-ca-file=/home/pgbackrest/certs/ca.crt
```

**Postgres server**
```
[global]
repo1-host=pgbackrest
repo1-host-user=pgbackrest
repo1-host-type=tls
repo1-host-cert-file=/var/lib/pgsql/certs/pg1-srv.crt
repo1-host-key-file=/var/lib/pgsql/certs/pg1-srv.key
repo1-host-ca-file=/var/lib/pgsql/certs/ca.crt
repo1-host-config=/tmp/pgbackrest.conf

# general options
process-max=2
log-level-console=info
log-level-file=debug

# tls server options
tls-server-address=*
tls-server-cert-file=/var/lib/pgsql/certs/pg1-srv.crt
tls-server-key-file=/var/lib/pgsql/certs/pg1-srv.key
tls-server-ca-file=/var/lib/pgsql/certs/ca.crt
tls-server-auth=pgbackrest=demo

[demo]
pg1-path=/var/lib/pgsql/14/data
```

**Comment by dwsteele** [comment]
Glad you got it working, but I want to mention (also for posterity) that changing the default config path is not ideal. It makes administering and debugging pgbackrest much harder and should be avoided if at all possible.
